### PR TITLE
Cleanup and reduce flakiness of some post-copy migration functests

### LINKED ISSUE #13435: Issue of postcopy migration with the check-tests-for-flakes lane
**What happened**:
During the PR https://github.com/kubevirt/kubevirt/pull/11833 an issue with the postcopy migration was seen on the check-test-for-flakes lane:
* https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11833/pull-kubevirt-check-tests-for-flakes/1861170200984948736
* https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11833/pull-kubevirt-check-tests-for-flakes/1861139610826969088

Another PR https://github.com/kubevirt/kubevirt/pull/13341 was opened to investigate the issue.

/kind flake
/priority critical-urgent
/sig compute

/assign @jean-edouard @fossedihelm

### PR DESCRIPTION
**Pull request by jean-edouard**
### What this PR does
Before this PR:
Post-copy tests are flaky

After this PR:
Code is cleaner and post-copy tests are less flaky

Fixes #13435

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer

Please review commit by commit!

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/migration/migration.go [hunk 1]
```diff
@@ -920,7 +920,7 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 
 				// Only stressing the VMI for 60 seconds to ensure the first migration eventually succeeds
 				By("Stressing the VMI")
-				runStressTest(vmi, stressDefaultVMSize, 60)
+				runStressTest(vmi, stressDefaultVMSize)
```
Comment (fossedihelm) [inline tests/migration/migration.go:923]:
```
// Only stressing the VMI for 60 seconds to ensure the first migration eventually succeeds
```
I see this comment... Don't we need this anymore? I mean this 60 seconds are not needed anymore?

File: tests/migration/postcopy.go [hunk 2]
```diff
@@ -168,39 +168,32 @@ var _ = SIGMigrationDescribe("VM Post Copy Live Migration", func() {
...
-					}, 5*time.Minute, 1*time.Second).Should(
-						SatisfyAll(HaveOccurred(), WithTransform(errors.IsNotFound, BeTrue())),
-						"The killer pod should be gone within the given timeout",
-					)
-				}
+				Eventually(func() error {
+					_, err := virtClient.CoreV1().Pods(testsuite.NamespacePrivileged).Get(context.Background(), killerPod, metav1.GetOptions{})
+					return err
+				}, 5*time.Minute, 1*time.Second).Should(
+					MatchError(errors.IsNotFound, "k8serrors.IsNotFound"), "The killer pod should be gone within the given timeout")
```
Comment (fossedihelm) [inline tests/migration/postcopy.go:196]: I am not sure this makes sense. If the Eventually block above succeded, it means that the killerPod is already gone

**Review by fossedihelm — COMMENTED** [review]
Looks great! Just two comments below. Thank you

File: tests/migration/migration.go (hunk 1 above)
Comment (jean-edouard) [reply tests/migration/migration.go:923]: I need to clean the comment, but if you look at the function it always reduced the timeout to 15 second max anyway...

File: tests/migration/postcopy.go (hunk 2 above)
Comment (jean-edouard) [reply tests/migration/postcopy.go:196]: I agree! This is untouched from the original version, I thought maybe there was some weird reason for it, but you're right I'll just nuke it.

File: tests/migration/migration.go (hunk 1 above)
Comment (fossedihelm) [reply tests/migration/migration.go:923]: lol xD

**Review by fossedihelm — APPROVED** [review]
Thank you!

**Comment by fossedihelm** [comment]
/lgtm
Fixes: https://github.com/kubevirt/kubevirt/issues/13435

File: tests/libmigration/migration.go [hunk 3]
```diff
@@ -500,16 +500,15 @@ func ConfirmMigrationMode(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMach
...
 }
 
-func WaitUntilMigrationMode(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, expectedMode v1.MigrationMode, timeout int) *v1.VirtualMachineInstance {
+func WaitUntilMigrationMode(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, expectedMode v1.MigrationMode, timeout time.Duration) {
 	By("Waiting until migration status")
 	EventuallyWithOffset(1, func() v1.MigrationMode {
-		By("Retrieving the VMI post migration")
-		vmi, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
+		By("Retrieving the VMI to check its migration mode")
+		newVMI, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
```
Comment (iholder101) [inline tests/libmigration/migration.go:507]: supernit: newVMI -> updatedVMI?

File: tests/migration/postcopy.go [hunk 4]
```diff
@@ -168,39 +168,32 @@ var _ = SIGMigrationDescribe("VM Post Copy Live Migration", func() {
...
-			var createdPods []string
-			BeforeEach(func() {
-				createdPods = []string{}
-			})
+			var killerPod string
 
-			runMigrationKillerPod := func(nodeName string) {
-				podName := fmt.Sprintf("migration-killer-pod-%s", rand.String(5))
+			runVirtHandlerKillerPod := func(nodeName string) {
+				podName := "migration-killer-pod-"
```
Comment (iholder101) [inline tests/migration/postcopy.go:174]: nit: maybe we want to call it (and the corresponding argument to `RenderPrivilegedPod()`) something like "name pattern"?

File: tests/migration/postcopy.go [hunk 5]
```diff
@@ -168,39 +168,32 @@ var _ = SIGMigrationDescribe("VM Post Copy Live Migration", func() {
...
-				for _, podName := range createdPods {
-					Eventually(func() error {
-						err := virtClient.CoreV1().Pods(testsuite.NamespacePrivileged).Delete(context.Background(), podName, metav1.DeleteOptions{})
-						return err
-					}, 10*time.Second, 1*time.Second).Should(MatchError(errors.IsNotFound, "k8serrors.IsNotFound"), "Should delete helper pod")
+			removeVirtHandlerKillerPod := func() {
+				Expect(killerPod).NotTo(BeEmpty())
+				Eventually(func() error {
+					err := virtClient.CoreV1().Pods(testsuite.NamespacePrivileged).Delete(context.Background(), killerPod, metav1.DeleteOptions{})
+					return err
```
Comment (iholder101) [inline tests/migration/postcopy.go:189]: supernit:
```suggestion
					return virtClient.CoreV1().Pods(testsuite.NamespacePrivileged).Delete(context.Background(), killerPod, metav1.DeleteOptions{})
```

File: tests/migration/postcopy.go [hunk 6]
```diff
@@ -208,6 +208,12 @@ var _ = SIGMigrationDescribe("VM Post Copy Live Migration", func() {
 					return fmt.Errorf("waiting for virt-handler pod to come back online")
 				}, 2*time.Minute, 1*time.Second).Should(Succeed(), "Virt handler should come online")
 			}
+
+			AfterEach(func() {
+				By("Ensuring the virt-handler killer pod is removed")
+				removeVirtHandlerKillerPod()
+			})
```
Comment (iholder101) [inline tests/migration/postcopy.go:209]: I think this is already taken care of by:
https://github.com/kubevirt/kubevirt/blob/9c9782a62bd63817b8103b0fa863b45c626b974f/tests/testsuite/namespace.go#L153-L154

**Review by iholder101 — COMMENTED** [review]
Looks great @jean-edouard, thank you!
Left mostly very minor comments.

**Comment by fossedihelm** [comment]
@jean-edouard can you add the `Fixes: https://github.com/kubevirt/kubevirt/issues/13435` in the PR description?
It still lgtm for me.
/lgtm

**Review by dhiller — COMMENTED** [review]
/approve

@iholder101 not sure whether you are good with this PR now?

Placing a 

/hold

just in case.

File: tests/migration/postcopy.go [hunk 7]
```diff
@@ -170,7 +170,7 @@ var _ = SIGMigrationDescribe("VM Post Copy Live Migration", func() {
 				podName := "migration-killer-pod-"
 
 				// kill the handler
-				pod := libpod.RenderPrivilegedPod(podName, []string{"/bin/bash", "-c"}, []string{fmt.Sprintf("while true; do pkill -9 virt-handler && sleep 5; done")})
+				pod := libpod.RenderPrivilegedPod(podName, []string{"/bin/bash", "-c"}, []string{"date; pkill -e -9 virt-handler || echo not found"})
```
Comment (xpivarc) [inline tests/migration/postcopy.go:173]: Why is this needed?

Comment (jean-edouard) [reply tests/migration/postcopy.go:173]: Which part? The main thing here is getting rid of the loop. Killing virt-handler once should be enough.
The date, `-e` and echo are to help debug failures. Before my change, the script is completely silent.

Comment (xpivarc) [reply tests/migration/postcopy.go:173]: The loop part, I remember I relied on it at least in one test. Please keep it if it is not a fix in which case please apply it only to specific tests.

Comment (jean-edouard) [reply tests/migration/postcopy.go:173]: This is used by only one test, and that test just needs virt-handler to disappear once, and in fact needs it to be there after to mark things as failed.
Maybe this function is copied somewhere else and that's what you remember?

Comment (xpivarc) [reply tests/migration/postcopy.go:173]: Correct, thanks!

**Comment by xpivarc** [comment]
/hold cancel
Let's get the flake fixes in and follow up if needed.
